devtools/src/arcs-hcr-list.js

```diff
@@ -0,0 +1,314 @@
+        display: inline-block;
+      }
+    </style>
+    <header class="header">
+      <div section>
+        <span class="devtools-icon refresh" on-click="_forceReload"></span>
+        <div divider></div>
+        <div class="tooltip">
+          <label class="switch">
+            <input type="checkbox" on-click="_loadServer" id="toggle">
+            <span class="slider round"></span>
+          </label>
+          <span class="tooltiptext">
+            Hot Code Reload
+          </span>
+        </div>
+      </div>
+    </header>
+
+    <div main>
+        <div class="title">Reloadable Particles</div>
+        <div class="content">
+          <template is="dom-repeat" items="{{reloadableParticles}}">
+            <object-explorer log-id$="[[item.implFile]]" object="{{item}}">
+              <span name>[[item.name]]</span>
+              <span tags>[[_tagsString(item.tags)]]</span>
+              <span type>[[_typeString(item.type)]]</span>
+              <span id>[[item.id]]</span>
+            </object-explorer>
+          </template>
+          <template is="dom-if" if="{{!item.items.length}}">
+            <div class="empty-label">No Particles</div>
+          </template>
+        </div>
+    </div>`;
+  }
+
+  static get is() { return 'arcs-hcr-list'; }
+
+  constructor() {
+    super();
+    this.websocket = null;
+    this.particleSources = new Set();
+    this.arcIdsToKeepFresh = [];
+    this.logging = new Map();
+    this.reloadableParticles = [];
+  }
+
+  onRawMessageBundle(messages) {
+    for (const msg of messages) {
+      switch (msg.messageType) {
+        case 'arc-available': {
+          if (!msg.messageBody.speculative) {
+            const id = msg.meta.arcId.toString();
+            this.arcIdsToKeepFresh.push(id);
+          }
+          break;
+        }
+        case 'watch-particle-sources': {
+          const before = this.particleSources.size;
+          for (const particleSource of msg.messageBody) {
+            this.particleSources.add(particleSource);
+          }
+          if (this.particleSources.size > before && this.websocket != null) {
+            const files = JSON.stringify([...this.particleSources]);
+            this.websocket.send(files);
+          } 
+          break;
+        }
+        case 'update-log': {
+          const filePath = msg.messageBody;
+          [...this.logging.values()].forEach(particles => {
+            particles.forEach(particle => {
+              if (particle.implFile == filePath) {
+                this._addLog(filePath, particle);
+                const explorer =  this.shadowRoot.querySelector(`object-explorer[log-id="${filePath}"]`);
+                explorer.refresh();
+                explorer.flash();
+              }
+            });
+          });
+          break;
+        }
+        case 'start-log': {
+          for (const particle of msg.messageBody) {
+            this._addLog(particle.implFile, particle, msg.meta.arcId.toString());
```

Comment:
It seems to me that the list of particles in an arc could be taken from the 'recipe-instantiated' messages, or alternatively from the 'InstantiateParticle' message type in the PecLog, depending on what exactly you need. Then you could build a UI list of particles based on these messages using only onMessageBundle(), they way the graph in arcs-overview rebuilds itself on switching between arcs.

Reply:
Added a listener to 'InstantiateParticle' message in PecLog and changed things to be much cleaner imo. Now the log only track `file -> time log` instead of `file -> particles`. Whenever arc selector is changed the list rebuilds all the particles from scratch just like in arcs-overview (then grab the reload log from the mapping based on its source file).

This eliminates the messages between devtools and runtime (don't need update-log and start-log anymore).